diff --git a/vendor/ospray-2.1.0/include/OSPEnums.h b/vendor/ospray-2.1.0/include/OSPEnums.h
index 804f554..8e70af2 100644
--- a/vendor/ospray-2.1.0/include/OSPEnums.h
+++ b/vendor/ospray-2.1.0/include/OSPEnums.h
@@ -8,9 +8,6 @@
 // Log levels which can be set on a driver via "logLevel" parameter or
 // "OSPRAY_LOG_LEVEL" environment variable
 typedef enum
-#if __cplusplus >= 201103L
-    : uint32_t
-#endif
 {
   OSP_LOG_DEBUG = 1,
   OSP_LOG_INFO = 2,
@@ -21,9 +18,6 @@ typedef enum
 
 // enum representing different device properties
 typedef enum
-#if __cplusplus >= 201103L
-    : uint32_t
-#endif
 {
   OSP_DEVICE_VERSION = 0,
   OSP_DEVICE_VERSION_MAJOR = 1,
@@ -34,9 +28,6 @@ typedef enum
 
 // An enum type that represensts the different data types represented in OSPRay
 typedef enum
-#if __cplusplus >= 201103L
-    : uint32_t
-#endif
 {
   // Object reference type.
   OSP_DEVICE = 100,
@@ -149,9 +140,6 @@ typedef enum
 
 // OSPRay format constants for Texture creation
 typedef enum
-#if __cplusplus >= 201103L
-    : uint32_t
-#endif
 {
   OSP_TEXTURE_RGBA8,
   OSP_TEXTURE_SRGBA,
@@ -174,9 +162,6 @@ typedef enum
 
 // Filter modes that can be set on 'texture2d' type OSPTexture
 typedef enum
-#if __cplusplus >= 201103L
-    : uint32_t
-#endif
 {
   OSP_TEXTURE_FILTER_BILINEAR = 0, // default bilinear interpolation
   OSP_TEXTURE_FILTER_NEAREST // use nearest-neighbor interpolation
@@ -184,9 +169,6 @@ typedef enum
 
 // Error codes returned by various API and callback functions
 typedef enum
-#if __cplusplus >= 201103L
-    : uint32_t
-#endif
 {
   OSP_NO_ERROR = 0, // No error has been recorded
   OSP_UNKNOWN_ERROR = 1, // An unknown error has occurred
@@ -203,9 +185,6 @@ typedef enum
 
 // OSPRay format constants for Frame Buffer creation
 typedef enum
-#if __cplusplus >= 201103L
-    : uint32_t
-#endif
 {
   OSP_FB_NONE, // framebuffer will not be mapped by application
   OSP_FB_RGBA8, // one dword per pixel: rgb+alpha, each one byte
@@ -216,9 +195,6 @@ typedef enum
 
 // OSPRay channel constants for Frame Buffer (can be OR'ed together)
 typedef enum
-#if __cplusplus >= 201103L
-    : uint32_t
-#endif
 {
   OSP_FB_COLOR = (1 << 0),
   OSP_FB_DEPTH = (1 << 1),
@@ -230,9 +206,6 @@ typedef enum
 
 // OSPRay events which can be waited on via ospWait()
 typedef enum
-#if __cplusplus >= 201103L
-    : uint32_t
-#endif
 {
   OSP_NONE_FINISHED = 0,
   OSP_WORLD_RENDERED = 10,
@@ -244,9 +217,6 @@ typedef enum
 // OSPRay cell types definition for unstructured volumes, values are set to
 // match VTK
 typedef enum
-#if __cplusplus >= 201103L
-    : uint8_t
-#endif
 {
   OSP_TETRAHEDRON = 10,
   OSP_HEXAHEDRON = 12,
@@ -257,9 +227,6 @@ typedef enum
 
 // OSPRay PerspectiveCamera stereo image modes
 typedef enum
-#if __cplusplus >= 201103L
-    : uint8_t
-#endif
 {
   OSP_STEREO_NONE,
   OSP_STEREO_LEFT,
@@ -270,9 +237,6 @@ typedef enum
 
 // OSPRay Curves geometry types
 typedef enum
-#if __cplusplus >= 201103L
-    : uint8_t
-#endif
 {
   OSP_ROUND,
   OSP_FLAT,
@@ -282,9 +246,6 @@ typedef enum
 
 // OSPRay Curves geometry bases
 typedef enum
-#if __cplusplus >= 201103L
-    : uint8_t
-#endif
 {
   OSP_LINEAR,
   OSP_BEZIER,
@@ -296,9 +257,6 @@ typedef enum
 
 // AMR Volume rendering methods
 typedef enum
-#if __cplusplus >= 201103L
-    : uint8_t
-#endif
 {
   OSP_AMR_CURRENT,
   OSP_AMR_FINEST,
@@ -307,9 +265,6 @@ typedef enum
 
 // Subdivision modes
 typedef enum
-#if __cplusplus >= 201103L
-    : uint8_t
-#endif
 {
   OSP_SUBDIVISION_NO_BOUNDARY,
   OSP_SUBDIVISION_SMOOTH_BOUNDARY,
diff --git a/vendor/ospray-2.1.0/include/ospray.h b/vendor/ospray-2.1.0/include/ospray.h
index dffa068..b63b295 100644
--- a/vendor/ospray-2.1.0/include/ospray.h
+++ b/vendor/ospray-2.1.0/include/ospray.h
@@ -3,44 +3,54 @@
 
 #pragma once
 
-#ifndef NULL
-#if __cplusplus >= 201103L
-#define NULL nullptr
-#else
-#define NULL 0
-#endif
-#endif
-
-#include <stdint.h>
-#include <sys/types.h>
-
-#include "OSPEnums.h"
-
-#ifdef _WIN32
-#ifdef ospray_EXPORTS
-#define OSPRAY_INTERFACE __declspec(dllexport)
-#else
-#define OSPRAY_INTERFACE __declspec(dllimport)
-#endif
-#else
-#define OSPRAY_INTERFACE
-#endif
-
-#ifdef __GNUC__
-#define OSP_DEPRECATED __attribute__((deprecated))
-#elif defined(_MSC_VER)
-#define OSP_DEPRECATED __declspec(deprecated)
-#else
-#define OSP_DEPRECATED
-#endif
-
-#ifdef __cplusplus
+#ifndef C2NIM
+    #ifndef NULL
+    #if __cplusplus >= 201103L
+    #define NULL nullptr
+    #else
+    #define NULL 0
+    #endif
+    #endif
+
+    #include <stdint.h>
+    #include <sys/types.h>
+    #include "OSPEnums.h"
+
+    #ifdef _WIN32
+    #ifdef ospray_EXPORTS
+    #define OSPRAY_INTERFACE __declspec(dllexport)
+    #else
+    #define OSPRAY_INTERFACE __declspec(dllimport)
+    #endif
+    #else
+    #define OSPRAY_INTERFACE
+    #endif
+
+    #ifdef __GNUC__
+        #define OSP_DEPRECATED __attribute__((deprecated))
+    #elif defined(_MSC_VER)
+        #define OSP_DEPRECATED __declspec(deprecated)
+    #else
+        #define OSP_DEPRECATED
+    #endif
+
+    #ifdef __cplusplus
 // C++ DOES support default initializers
-#define OSP_DEFAULT_VAL(a) = a
-#else
+        #define OSP_DEFAULT_VAL(a) = a
+    #else
 /* C99 does NOT support default initializers, so we use this macro
    to define them away */
-#define OSP_DEFAULT_VAL(a)
+        #define OSP_DEFAULT_VAL(a)
+    #endif
+#else
+    #dynlib ospraydll
+    #if defined(windows)
+      #define ospraydll "ospray.dll"
+    #else
+      #define ospraydll "libospray.so"
+    #endif
+    #def OSP_DEFAULT_VAL(a)
+    #def OSPRAY_INTERFACE 
 #endif
 
 // Give OSPRay handle types a concrete defintion to enable C++ type checking
@@ -131,10 +141,42 @@ typedef void _OSPManagedObject;
    OSPGeometry can still be passed to a function that expects a
    OSPObject, etc. */
 typedef _OSPManagedObject *OSPManagedObject, *OSPCamera, *OSPData,
-    *OSPFrameBuffer, *OSPFuture, *OSPGeometricModel, *OSPGeometry, *OSPGroup,
-    *OSPImageOperation, *OSPInstance, *OSPLight, *OSPMaterial, *OSPObject,
-    *OSPRenderer, *OSPTexture, *OSPTransferFunction, *OSPVolume,
-    *OSPVolumetricModel, *OSPWorld;
+*OSPFrameBuffer, *OSPFuture, *OSPGeometricModel, *OSPGeometry, *OSPGroup,
+*OSPImageOperation, *OSPInstance, *OSPLight, *OSPMaterial, *OSPObject,
+*OSPRenderer, *OSPTexture, *OSPTransferFunction, *OSPVolume,
+*OSPVolumetricModel, *OSPWorld;
+#endif
+
+#ifdef C2NIM
+#@
+import OSPEnums
+export OSPEnums
+type
+    OSPDevice* = distinct pointer
+    OSPManagedObject* = object of RootObj
+    OSPCamera* = object of OSPManagedObject
+    OSPData* = object of OSPManagedObject
+    OSPFrameBuffer* = object of OSPManagedObject
+    OSPFuture* = object of OSPManagedObject
+    OSPGeometricModel* = object of OSPManagedObject
+    OSPGeometry* = object of OSPManagedObject
+    OSPGroup* = object of OSPManagedObject
+    OSPImageOperation* = object of OSPManagedObject
+    OSPInstance* = object of OSPManagedObject
+    OSPLight* = object of OSPManagedObject
+    OSPMaterial* = object of OSPManagedObject
+    OSPObject* = object of OSPManagedObject
+    OSPRenderer* = object of OSPManagedObject
+    OSPTexture* = object of OSPManagedObject
+    OSPTransferFunction* = object of OSPManagedObject
+    OSPVolume* = object of OSPManagedObject
+    OSPVolumetricModel* = object of OSPManagedObject
+    OSPWorld* = object of OSPManagedObject
+    int64_t* = int64
+    uint64_t* = uint64
+    uint32_t* = uint32
+    int32_t* = int32
+@#
 #endif
 
 #ifdef __cplusplus
diff --git a/vendor/ospray-2.1.0/include/ospray_util.h b/vendor/ospray-2.1.0/include/ospray_util.h
index 98f689a..c328c1b 100644
--- a/vendor/ospray-2.1.0/include/ospray_util.h
+++ b/vendor/ospray-2.1.0/include/ospray_util.h
@@ -5,6 +5,17 @@
 
 #include "ospray.h"
 
+#ifdef C2NIM
+    #dynlib ospraydll
+    #if defined(windows)
+      #define ospraydll "ospray.dll"
+    #else
+      #define ospraydll "libospray.so"
+    #endif
+    #def OSP_DEFAULT_VAL(a)
+    #def OSPRAY_INTERFACE 
+#endif
+
 #ifdef __cplusplus
 extern "C" {
 #endif
